src/custom/state/enhancedTransactions/updater.tsx

```diff
@@ -0,0 +1,52 @@
+import { useEffect } from 'react'
+import { useAppDispatch } from 'state/hooks'
+import { useActiveWeb3React } from 'hooks/web3'
+import { sdk } from 'utils/blocknative'
+import { cancelTransaction, replaceTransaction } from 'state/enhancedTransactions/actions'
+import { useAllPendingHashes } from 'state/enhancedTransactions/hooks'
+
+export default function Updater(): null {
+  const { chainId, library } = useActiveWeb3React()
+  const dispatch = useAppDispatch()
+  const pendingHashes = useAllPendingHashes()
+
+  useEffect(() => {
+    if (!chainId || !library) return
+
+    for (const hash of pendingHashes) {
+      try {
+        const { emitter } = sdk[chainId].transaction(hash)
+        const currentHash = hash
+        let isSpeedup = false
+
+        emitter.on('txSpeedUp', (e) => {
+          isSpeedup = true
+          if ('hash' in e && typeof e.hash === 'string') {
+            dispatch(replaceTransaction({ chainId, oldHash: currentHash, newHash: e.hash }))
+          }
+        })
+
+        emitter.on('txConfirmed', (e) => {
+          // canceled txs are automatically watched by bnc so if the confirmed tx hash is different than the previously tracked one, it means the user sent a cancel tx
+          if ('hash' in e && e.hash !== currentHash && !isSpeedup) {
+            dispatch(cancelTransaction({ chainId, hash: currentHash }))
+          }
+        })
+      } catch (error) {
```

Comment:
Excellent work @Ramiro, you convinced me with bn.

Just a doub here. What would happen with this transaction hash that was neither replaced nor cancelled, would it be there as pending until the application is restarted?

Reply:
If a pending transaction is watched and then that transaction is confirmed, the pending hashes array will change and that will cause the useEffect cleanup function to kick in, unsubscribing any previous watched hash.